stdlib/typing.pyi

```diff
@@ -563,19 +564,29 @@ class Match(Generic[AnyStr]):
     # this match instance.
     re: Pattern[AnyStr]
     def expand(self, template: AnyStr) -> AnyStr: ...
-    # TODO: The return for a group may be None, except if __group is 0 or not given.
+    # group() returns "AnyStr" or "AnyStr | None", depending on the pattern.
     @overload
-    def group(self, __group: Union[str, int] = ...) -> AnyStr: ...
+    def group(self, __group: _Literal[0] = ...) -> AnyStr: ...
     @overload
-    def group(self, __group1: Union[str, int], __group2: Union[str, int], *groups: Union[str, int]) -> Tuple[AnyStr, ...]: ...
-    def groups(self, default: AnyStr = ...) -> Sequence[AnyStr]: ...
-    def groupdict(self, default: AnyStr = ...) -> dict[str, AnyStr]: ...
+    def group(self, __group: str | int) -> AnyStr | Any: ...
+    @overload
+    def group(self, __group1: str | int, __group2: str | int, *groups: str | int) -> Tuple[AnyStr | Any, ...]: ...
+    # Each item of groups()'s return tuple is either "AnyStr" or
+    # "AnyStr | None", depending on the pattern.
+    def groups(self, default: AnyStr = ...) -> Tuple[AnyStr | Any, ...]: ...
```

Comment:
The `default` annotation is wrong:
- If `default` is None (the default), then the resulting tuple can contain `None`s.
- If `default` is `AnyStr`, then the resulting tuple will not contain `None`s.

```
>>> print(re.fullmatch('a(b)?', 'a').groups())
(None,)
>>> print(re.fullmatch('a(b)?', 'a').groups(default=None))
(None,)
>>> print(re.fullmatch('a(b)?', 'a').groups(default="foo"))
('foo',)
>>> 
```

The default can in fact be any object, so maybe this with a TypeVar named `_T`?

```suggestion
    @overload
    def groups(self, default: None = ...) -> Tuple[AnyStr | Any, ...]: ...
    @overload
    def groups(self, default: _T) -> Tuple[AnyStr | _T, ...]: ...
```

This also applies to `.groupdict()`.

Reply:
I actually changed the first overload not to accept any argument, with the rationale that if someone uses `m.groups(default=None)`, they are prepared to deal with `None` values in the result tuple.